<div style="border:solid Chocolate 2px; padding: 40px">

<b> Виталий, привет!👋</b>

Меня зовут Алексей Гриб, и я буду ревьюером твоего проекта. 

Сразу хочу предложить в дальнейшем общаться на "ты" - надеюсь, так будет комфортнее:) Но если это неудобно, обязательно дай знать, и мы придумаем что-нибудь ещё!
    
Цель ревью - не искать ошибки в твоём проекте, а помочь тебе сделать твою работу ещё лучше, устранив недочёты и приблизив её к реальным задачам специалиста по работе с данными. Поэтому не расстраивайся, если что-то не получилось с первого раза - это нормально, и это поможет тебе вырасти!
    
Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> похвала, рекомендации «со звёздочкой», полезные лайфхаки, которые сделают и без того красивое решение ещё более элегантным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> некритичные ошибки или развивающие рекомендации на будущее. 
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
критичные ошибки, которые обязательно нужно исправить.
</div>

    
Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 
    
Ты также можешь задавать свои вопросы, реагировать на мои комментарии, делать пометки и пояснения - полная творческая свобода! Но маленькая просьба - пускай они будут отличаться от моих комментариев, это поможет избежать путаницы в нашем общении:)
Например, вот так:
    
<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*твой текст*
</div>
    
Давай посмотрим на твой проект!

# Исследовательский анализ данных Яндекс Афиши

## Цели проекта

- Выявить причины изменений в пользовательском спросе в сервисе «Яндекс Афиша».
- Определить наиболее востребованные категории событий, партнёров и регионов.
- Оценить влияние сезонности на активность пользователей и популярность мероприятий.
- Сравнить поведение пользователей мобильных устройств и стационарных компьютеров.
- Проверить гипотезы об активности пользователей на разных типах устройств.
- Сформировать рекомендации для продуктовой команды.

## Структура проекта

### **Шаг 0. Оформление**
- Название, цели, структура, описание данных.

### **Шаг 1. Загрузка и первичный анализ данных**
- Загрузка данных, первичная проверка корректности, объёма и описания.
- Определение необходимых шагов предобработки.

### **Шаг 2. Предобработка данных**
- Работа с пропусками, дубликатами, аномалиями.
- Приведение выручки к единой валюте — рублям.
- Добавление новых признаков: `revenue_rub`, `one_ticket_revenue_rub`, `month`, `season`.

### **Шаг 3. Исследовательский анализ данных**

#### 3.1 Анализ сезонности и сегментов
- Динамика заказов по месяцам и сезонам.
- Сравнение распределения заказов летом и осенью:
  - по типу мероприятия;
  - по типу устройства;
  - по возрастным ограничениям.
- Анализ изменения средней выручки с одного билета по категориям.

#### 3.2 Осенняя активность пользователей
- Динамика по дням: число заказов, DAU, среднее число заказов на пользователя, стоимость билета.
- Анализ цикличности (будни vs выходные).

#### 3.3 Популярные события и партнёры
- Популярные регионы и партнёры: количество мероприятий, заказов, выручка.
- Расчёт долей и сравнение распределения.

### **Шаг 4. Статистический анализ**

- Проверка двух гипотез:
  1. Среднее число заказов на пользователя выше у мобильных пользователей.
  2. Среднее время между заказами больше у мобильных пользователей.
- Обоснование выбора теста, формулировка H₀ и H₁, анализ результатов.

### **Шаг 5. Общий вывод и рекомендации**
- Краткие итоги исследования.
- Ответы на ключевые вопросы.
- Рекомендации для команды Яндекс Афиши.

## Описание данных проекта
### Первый датасет final_tickets_orders_df
- order_id — уникальный идентификатор заказа.
- user_id — уникальный идентификатор пользователя.
- created_dt_msk — дата создания заказа (московское время).
- created_ts_msk — дата и время создания заказа (московское время).
- event_id — идентификатор мероприятия из таблицы events.
- cinema_circuit — сеть кинотеатров. Если не применимо, то здесь будет значение 'нет'.
- age_limit — возрастное ограничение мероприятия.
- currency_code — валюта оплаты, например rub для российских рублей.
- device_type_canonical — тип устройства, с которого был оформлен заказ, например mobile для мобильных устройств, desktop для стационарных.
- revenue — выручка от заказа.
- service_name — название билетного оператора.
- tickets_count — количество купленных билетов.

### Второй датасет final_tickets_events_df
- event_id — уникальный идентификатор мероприятия.
- event_name — название мероприятия. Аналог поля event_name_code из исходной базы данных.
- event_type_description — описание типа мероприятия.
- event_type_main — основной тип мероприятия: театральная постановка, концерт и так далее.
- organizers — организаторы мероприятия.
- region_name — название региона.
- city_name — название города.
- venue_id — уникальный идентификатор площадки.
- venue_name — название площадки.
- venue_address — адрес площадки.

### Дополнительный датасет final_tickets_tenge_df
- nominal — номинал (100 тенге).
- data — дата.
- curs — курс тенге к рублю.
- cdx — обозначение валюты (kzt).

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
<b>Все отлично!👍:</b> Хорошее вступление!
    
В нём есть всё, что необходимо, чтобы понять суть проекта с первых строк отчёта!

## Шаг 1. Загрузка данных и знакомство с ними

In [1]:
!pip install numpy -U
!pip install matplotlib -U
!pip install scipy -U
!pip install pandas -U
!pip install seaborn -U

zsh:1: command not found: pip
zsh:1: command not found: pip
zsh:1: command not found: pip
zsh:1: command not found: pip
zsh:1: command not found: pip


In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from scipy.stats import mannwhitneyu, ks_2samp

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Библиотеки импортировали - отлично!

In [3]:
df_orders = pd.read_csv('https://code.s3.yandex.net//datasets/final_tickets_orders_df.csv')
df_events = pd.read_csv('https://code.s3.yandex.net//datasets/final_tickets_events_df.csv')
df_currency = pd.read_csv('https://code.s3.yandex.net//datasets/final_tickets_tenge_df.csv')

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> 
    
Данные загрузили.
    
При считывании данных из файла здорово перестраховывать себя от ошибок, связанных, например, с неверным указанием пути к файлу. А иногда бывает, что работаешь с файлом локально, выгружаешь его на сервер, ожидая, что он будет принимать данные, которые лежат на том же сервере, а код падает с ошибкой, потому что путь к файлу не поменялся с локального на серверный.
    
Для этого, например, можно использовать конструкцию `try-except`: сначала пробуешь локальный путь, при возникновении ошибки используется серверный путь (подробнее можешь почитать тут: https://pythonworld.ru/tipy-dannyx-v-python/isklyucheniya-v-python-konstrukciya-try-except-dlya-obrabotki-isklyuchenij.html).
    
Но еще лучше использовать библиотеку `os` - её использование позволит тебе проверять существование указанных директорий (что может быть актуально при одновременной работа на локальном и сетевом окружении) и загружать данные из существующей директории, избегая ошибок. Как пример:
    
    import os

    pth1 = '/folder_1/data.csv'
    pth2 = '/folder_2/data.csv'
    
    if os.path.exists(pth1):
        query_1 = pd.read_csv(pth1)
    elif os.path.exists(pth2):
        query_1 = pd.read_csv(pth2)
    else:
        print('Something is wrong')
    
Обрати внимание, что при считывании данных по `url` адресу (вроде `https://code.s3.yandex.net//datasets/`), `os` не сможет их проверить, так как работает только с физическими путями. В этом случае можно сделать проверку через `requests`: можно отправить `.get()` запрос и проверить `status_code` в ответе. Если он будет `200`, данные можно загружать.

Ещё на этапе считывания данных можно спарсить дату: за это действие отвечает параметр `parse_dates` метода `read_csv()`, в него нужно передать список с названием полей-дат, и в большинстве случаев дата будет корректно преобразована в нужный формат сразу:)
Также на этапе считывания данных можно задать индекс-столбец - за это действие отвечает параметр `index_col`.

In [4]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 290849 entries, 0 to 290848
Data columns (total 14 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   order_id               290849 non-null  int64  
 1   user_id                290849 non-null  object 
 2   created_dt_msk         290849 non-null  object 
 3   created_ts_msk         290849 non-null  object 
 4   event_id               290849 non-null  int64  
 5   cinema_circuit         290849 non-null  object 
 6   age_limit              290849 non-null  int64  
 7   currency_code          290849 non-null  object 
 8   device_type_canonical  290849 non-null  object 
 9   revenue                290849 non-null  float64
 10  service_name           290849 non-null  object 
 11  tickets_count          290849 non-null  int64  
 12  total                  290849 non-null  float64
 13  days_since_prev        268909 non-null  float64
dtypes: float64(3), int64(4), object(7)
m

In [5]:
df_orders.head()

order_id          user_id created_dt_msk       created_ts_msk  event_id  \
0   4359165  0002849b70a3ce2     2024-08-20  2024-08-20 16:08:03    169230   
1   7965605  0005ca5e93f2cf4     2024-07-23  2024-07-23 18:36:24    237325   
2   7292370  0005ca5e93f2cf4     2024-10-06  2024-10-06 13:56:02    578454   
3   1139875  000898990054619     2024-07-13  2024-07-13 19:40:48    387271   
4    972400  000898990054619     2024-10-04  2024-10-04 22:33:15    509453   

  cinema_circuit  age_limit currency_code device_type_canonical  revenue  \
0            нет         16           rub                mobile  1521.94   
1            нет          0           rub                mobile   289.45   
2            нет          0           rub                mobile  1258.57   
3            нет          0           rub                mobile     8.49   
4            нет         18           rub                mobile  1390.41   

         service_name  tickets_count     total  days_since_prev  
0        Край билетов              4  10870.99              NaN  
1           Мой билет              2   2067.51              NaN  
2         За билетом!              4  13984.16             75.0  
3         Лови билет!              2    212.28              NaN  
4  Билеты без проблем              3  10695.43             83.0

In [6]:
df_events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22427 entries, 0 to 22426
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   event_id                22427 non-null  int64 
 1   event_name              22427 non-null  object
 2   event_type_description  22427 non-null  object
 3   event_type_main         22427 non-null  object
 4   organizers              22427 non-null  object
 5   region_name             22427 non-null  object
 6   city_name               22427 non-null  object
 7   city_id                 22427 non-null  int64 
 8   venue_id                22427 non-null  int64 
 9   venue_name              22427 non-null  object
 10  venue_address           22427 non-null  object
dtypes: int64(3), object(8)
memory usage: 1.9+ MB


In [7]:
df_events.head()

event_id                            event_name event_type_description  \
0      4436  e4f26fba-da77-4c61-928a-6c3e434d793f              спектакль   
1      5785  5cc08a60-fdea-4186-9bb2-bffc3603fb77              спектакль   
2      8817  8e379a89-3a10-4811-ba06-ec22ebebe989              спектакль   
3      8849  682e3129-6a32-4952-9d8a-ef7f60d4c247              спектакль   
4      8850  d6e99176-c77f-4af0-9222-07c571f6c624              спектакль   

  event_type_main organizers           region_name   city_name  city_id  \
0           театр      №4893  Североярская область      Озёрск        2   
1           театр      №1931  Светополянский округ   Глиноград       54   
2           театр      №4896  Североярская область      Озёрск        2   
3           театр      №4960    Каменевский регион  Глиногорск      213   
4           театр      №4770      Лесодальний край  Родниковец       55   

   venue_id                               venue_name             venue_address  
0      1600  Кладбище искусств "Проблема" и партнеры   наб. Загородная, д. 785  
1      2196         Лекции по искусству "Свет" Групп       ул. Ягодная, д. 942  
2      4043                Кинокомитет "Золотая" Инк  ш. Коммуны, д. 92 стр. 6  
3      1987              Выставка ремесел "Свет" Лтд    пер. Набережный, д. 35  
4      4230       Фестивальный проект "Листья" Групп       пер. Проезжий, д. 9

In [8]:
df_currency.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 357 entries, 0 to 356
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   data     357 non-null    object 
 1   nominal  357 non-null    int64  
 2   curs     357 non-null    float64
 3   cdx      357 non-null    object 
dtypes: float64(1), int64(1), object(2)
memory usage: 11.3+ KB


In [9]:
df_currency.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 357 entries, 0 to 356
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   data     357 non-null    object 
 1   nominal  357 non-null    int64  
 2   curs     357 non-null    float64
 3   cdx      357 non-null    object 
dtypes: float64(1), int64(1), object(2)
memory usage: 11.3+ KB


In [10]:
df_currency.head()

data  nominal     curs  cdx
0  2024-01-10      100  19.9391  kzt
1  2024-01-11      100  19.7255  kzt
2  2024-01-12      100  19.5839  kzt
3  2024-01-13      100  19.4501  kzt
4  2024-01-14      100  19.4501  kzt

In [11]:
df_currency['nominal'].unique()

array([100])

### Промежуточные выводы после первичного анализа данных
1. Обнаружены пропуски в столбце days_since_prev, что вполне ожидаемо для первых покупок пользователей. Мероприятия и курсы валют пропусков не содержат. Все датасеты соответствуют описанию полей.
2. В качестве преобработки надо изменить тип данных в полях created_dt_msk, created_ts_msk в датафрейме df_orders и в поле data в датафрейме data_currency 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Первичный анализ данных выполнен, отмечено наличие пропусков и некорректных типов данных с датами.

# Шаг 2. Предобработка данных и подготовка их к исследованию

Проверяем данные на пропущенные значения 

In [12]:
print(df_orders.isnull().sum())

order_id                     0
user_id                      0
created_dt_msk               0
created_ts_msk               0
event_id                     0
cinema_circuit               0
age_limit                    0
currency_code                0
device_type_canonical        0
revenue                      0
service_name                 0
tickets_count                0
total                        0
days_since_prev          21940
dtype: int64


In [13]:
print(df_events.isnull().sum())

event_id                  0
event_name                0
event_type_description    0
event_type_main           0
organizers                0
region_name               0
city_name                 0
city_id                   0
venue_id                  0
venue_name                0
venue_address             0
dtype: int64


In [14]:
print(df_currency.isnull().sum())

data       0
nominal    0
curs       0
cdx        0
dtype: int64


Видим, что пропуски данных есть только в поле days_since_prev

Рассмотрим категориальные значения

In [20]:
print("\nУникальные значения device_type_canonical:")
print(df_orders['device_type_canonical'].value_counts())

print("\nУникальные значения currency_code:")
print(df_orders['currency_code'].value_counts())

print("\nУникальные значения cinema_circuit:")
print(df_orders['cinema_circuit'].value_counts())

print("\nУникальные значения event_type_main:")
print(df_events['event_type_main'].value_counts())

print("\nУникальные значения age_limit:")
print(df_orders['age_limit'].value_counts())

print("\nУникальные значения region_name:")
print(df_events['region_name'].value_counts())

print("\nУникальные значения service_name:")
print(df_orders['service_name'].value_counts())


Уникальные значения device_type_canonical:
device_type_canonical
mobile     232679
desktop     58170
Name: count, dtype: int64

Уникальные значения currency_code:
currency_code
rub    285780
kzt      5069
Name: count, dtype: int64

Уникальные значения cinema_circuit:
cinema_circuit
нет           289451
Другое          1261
КиноСити         122
Киномакс           7
Москино            7
ЦентрФильм         1
Name: count, dtype: int64

Уникальные значения event_type_main:
event_type_main
концерты    8680
театр       7076
другое      4658
спорт        872
стендап      636
выставки     290
ёлки         215
Name: count, dtype: int64

Уникальные значения age_limit:
age_limit
16    78579
12    62557
0     61487
6     52173
18    36053
Name: count, dtype: int64

Уникальные значения region_name:
region_name
Каменевский регион          5983
Североярская область        3814
Широковская область         1233
Светополянский округ        1075
Речиновская область          702
                          

По данным уникальных значений можно сделать следующие выводы: категория Mobile доминирует над категорией Desktop, основной валютой являются рубли, также самыми популярными категориями являются концерты и театр.

### Выводы по предобработке данных и подготовке к исследованию

#### 1. Пропущенные значения

- Датасет заказов (orders): 
  В столбце days_since_prev присутствуют 21 940 пропусков. Скорее всего, они соответствуют первым заказам пользователей, для которых нет предыдущих покупок. Остальные поля не содержат пропусков.

- Датасет мероприятий (events):  
  Данные полные, пропущенных значений нет — можно использовать без дополнительной очистки.

- Датасет курса тенге (tenge):  
  Пропуски отсутствуют — таблица готова к дальнейшей работе.

#### 2. Анализ категориальных признаков

- Тип устройства (device_type_canonical):  
  Большинство заказов оформлено с мобильных устройств — 232 679 записей против 58 170 с десктопов, что соответствует современным трендам в e-commerce.

- Валюта (currency_code):  
  Основной валютой является российский рубль (RUB) — 285 780 заказов. Транзакции в тенге (KZT) встречаются значительно реже — 5 069.

- Киносеть (cinema_circuit): 
  В большинстве случаев (289 451 запись) заказ не относится к киносети. Среди упомянутых киносетей чаще всего встречается категория "Другое" (1 261), тогда как конкретные сети (КиноСити, Киномакс, Москино, ЦентрФильм) представлены в небольшом объёме.

- Тип мероприятия (event_type_main):
  Наиболее популярные категории — концерты (8 680) и театр (7 076). Менее распространены спорт (872), стендап (636), выставки (290) и ёлки (215).


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Категориальные признаки изучены.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
    
В части анализа категориальных признаков важно сделать ацент на наличии в их значениях каких-то аномалий - скрытых пропусков/дубликатов/ошибок. Это важно отметить в выводе. 
    
Также нужно проанализировать все категориальные признаки, задействованные в проекте.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Тут актуально - добавь, пожалуйста, в исследование все категориальные признаки, которые задействованы в главе 2.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

я сам идейно вижу такое добавление выводов
</div>

### Анализ категориальных данных


#### 1. device_type_canonical — Тип устройства
| Устройство | Количество | Доля |
|------------|------------|------|
| mobile     | 232,679    | ≈ 80% |
| desktop    | 58,170     | ≈ 20% |

Вывод:
Преобладают мобильные пользователи. Необходим приоритет mobile-first дизайна.


### 2. currency_code — Валюта
| Валюта | Количество | Доля |
|--------|------------|------|
| rub    | 285,780    | ≈ 98.3% |
| kzt    | 5,069      | ≈ 1.7%  |

Вывод:  
Целевая аудитория — в основном Россия. Казахстан — вторичный, но важный рынок.


### 3. cinema_circuit — Сеть кинотеатров
| Сеть       | Количество |
|------------|------------|
| нет        | 289,451    |
| Другое     | 1,261      |
| КиноСити   | 122        |
| Киномакс   | 7          |
| Москвино   | 7          |
| ЦентрФильм | 1          |

Вывод:
Почти все записи не привязаны к сетям. Возможна неполная заполняемость поля.


### 4. event_type_main — Основной тип события
| Тип события | Кол-во |
|-------------|--------|
| концерты    | 8,680  |
| театр       | 7,076  |
| другое      | 4,658  |
| спорт       | 872    |
| стендап     | 636    |
| выставки    | 290    |
| ёлки        | 215    |

Вывод: 
Концерты и театр — самые популярные категории. Высокая доля "другое" может указывать на необходимость пересмотра структуры классификации.


### 5. age_limit — Возрастное ограничение
| Возраст | Кол-во |
|---------|--------|
| 16      | 78,579 |
| 12      | 62,557 |
| 0       | 61,487 |
| 6       | 52,173 |
| 18      | 36,053 |

Вывод:
- Наиболее частые значения: 16, 12 и 0.  
- Наличие значения 0 может означать "без ограничений", но его следует явно описать или заменить.
- Возраст 18 встречается значительно реже, что может говорить о меньшей популярности мероприятий "только для взрослых".


### 6. region_name — Регион
- Всего уникальных регионов: 81
- Дисбаланс в частотах: от 5983 до 2 записей на регион

Вывод:
- Сильный дисбаланс: верхние 5 регионов дают значительную часть данных.
- Нижние регионы с 2–10 записями могут рассматриваться как выбросы или данные с низкой репрезентативностью.
- Возможна агрегация малых регионов в одну категорию типа "прочее" для анализа.


### 7. service_name — Название билетного сервиса
- Всего уникальных сервисов: 34
- Наиболее популярные:
  | Сервис              | Кол-во |
  |---------------------|--------|
  | Билеты без проблем  | 63,709 |
  | Лови билет!         | 41,126 |
  | Билеты в руки       | 40,364 |
  | Мой билет           | 34,843 |

- Наименее популярные (возможно выбросы или тестовые данные):
  | Сервис                 | Кол-во |
  |------------------------|--------|
  | Билеты в интернете     | 4      |
  | Зе Бест!               | 5      |
  | Лимоны                 | 8      |

Вывод: 
- Сервисы с крайне малым количеством записей могут быть:
  - Тестовыми / временными
  - Введёнными вручную
  - Ошибочными (особенно с нестандартными названиями)
- Для анализа можно исключить сервисы с частотой < 50 как потенциальные выбросы.


### Общие рекомендации:
1. Очистка и стандартизация данных:
   - Обработать возраст 0 (явно указать как "без ограничений")
   - Объединить редкие регионы в `"другие"`
   - Пересмотреть сервисы с низкой частотой — потенциально удалить или агрегировать

2. Снижение доли "другое" в типах событий за счёт уточнения классификации.

3. Фокус на мобильную версию интерфейса — подавляющее большинство пользователей с мобильных устройств.

4. Выбросы:
   - Регионы и сервисы с единичными значениями можно отнести к выбросам.
   - Рекомендуется пороговое отсечение или группировка в "прочее".

5. Стратегии сегментации:
   - По возрасту: возможно таргетирование на 12+ и 16+
   - По региону: приоритизация регионов с наибольшей активностью
   - По сервисам: анализировать топ-5 сервисов отдельно как ключевые каналы



In [ ]:
rub_orders = df_orders[df_orders['currency_code'] == 'rub']
kzt_orders = df_orders[df_orders['currency_code'] == 'kzt']

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> 

Делая копию или срез какой-то таблицы, стоит использовать метод `copy()`: используя простое присвивание вида `a=b`, мы в переменную `a` помещаем не новый и самостоятельный объект в памяти, а всего лишь новую ссылку на объект в памяти, на который ссылается `b`. Как итог, внеся изменения в `a`, ты увидишь их и в `b`, и наоборот. Метод `copy()` позволяет создавать именно новые объекты в памяти.

In [ ]:
plt.figure(figsize=(12, 6))
plt.subplot(1, 2, 1)
sns.boxplot(data=rub_orders, y='revenue')
plt.title('Выручка в рублях')

plt.subplot(1, 2, 2)
sns.boxplot(data=kzt_orders, y='revenue')
plt.title('Выручка в тенге')
plt.show()

revenue_rub_threshold = rub_orders['revenue'].quantile(0.99)
revenue_kzt_threshold = kzt_orders['revenue'].quantile(0.99)

df_orders = df_orders[~((df_orders['currency_code'] == 'rub') & (df_orders['revenue'] > revenue_rub_threshold) | 
               ((df_orders['currency_code'] == 'kzt') & (df_orders['revenue'] > revenue_kzt_threshold)))] #фильтруем выбросы в данных

Рассмотрим отдельно анализ числа билетов

In [ ]:
df_orders['tickets_count'].describe()

Проведем так же фильтрацию из-за выбросов

In [ ]:
tickets_threshold = df_orders['tickets_count'].quantile(0.99)  
df_orders = df_orders[df_orders['tickets_count'] <= tickets_threshold]

Можем сделать вывод, что максимальное значение билетов равно 57 и является выбросом, либо это был групповой заказ, либо ошибка
Типичный заказ это 3 билета.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Численные признаки изучены, удалены выбросы по числу билетов - верное решение.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Явные выбросы есть также в выручке по рублям - их также нужно обработать. В тенге можно обрабатывать опционально, так как выбросов немного. При этом обрати внимание, что обработку важно делать в разрезе валют.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>
Они обработаны уже, просто обработка находится в той части кода, где и вывод графиков
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Понял, спасибо, что уточнил - по тексту вывода ожидал увидеть фильтрацию тут.

Теперь проверим данные в заказах на явные и неявные дубликаты

In [ ]:
print("Явные дубликаты в заказах:", df_orders.duplicated().sum())

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Явные дубликаты проверены.

In [ ]:
duplicate_columns = ['user_id', 'created_dt_msk', 'event_id', 'total']
print("\nНеявные дубликаты:", df_orders.duplicated(subset=duplicate_columns).sum())

Видим, что явные дубликаты не обнаружены

In [ ]:
columns_to_check = df_orders.columns.drop('order_id')  
duplicates_all = df_orders.duplicated(subset=columns_to_check).sum()
print(f"Неявные дубликаты (все поля кроме order_id): {duplicates_all}") # мы проверяем неявные дубликаты по всем полям, кроме 

In [ ]:
df_orders_cleaned = df_orders.drop_duplicates(subset=columns_to_check, keep='first')

print(f"Удалено записей: {len(df_orders) - len(df_orders_cleaned)}")

Были найдены неявные дубликаты
Было принято решение удалить дубликаты с сохранением первой строчки, так как сохранение всех имеющихся записей могло исказить анализ

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Неявные дубликаты также в порядке.

Теперь нужно привести данные к нужному типу

In [ ]:
df_orders_cleaned = df_orders_cleaned.copy()  
df_orders_cleaned['created_dt_msk'] = pd.to_datetime(df_orders_cleaned['created_dt_msk'])
df_orders_cleaned['created_ts_msk'] = pd.to_datetime(df_orders_cleaned['created_ts_msk'])
df_orders_cleaned['tickets_count'] = pd.to_numeric(df_orders_cleaned['tickets_count'], downcast='integer')
df_orders_cleaned['age_limit'] = pd.to_numeric(df_orders_cleaned['age_limit'], downcast='integer')
df_currency['data'] = pd.to_datetime(df_currency['data'])

In [ ]:
print(df_orders_cleaned.dtypes[['created_dt_msk', 'created_ts_msk']])

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Типы данных скорректированы.

In [ ]:
df = pd.merge(df_orders_cleaned, df_events, on='event_id', how='left')

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Данные соединены корректно.

In [ ]:
df.head()

In [ ]:
df.info()

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> После соединения данных стоит проверить их на наличие новых пропусков.

In [ ]:
tenge_dict = {k.date(): v/100 for k, v in df_currency.set_index('data')['curs'].to_dict().items()}

def convert_to_rub(row):
    if row['currency_code'] == 'kzt':
        date = row['created_dt_msk'].date()
        rate = tenge_dict[date]  
        return row['revenue'] * rate
    return row['revenue']

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> Достаточно тяжеловесное решение - стоит попробовать делать присваивания через `.loc` по условию нужной валюты.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Учет деления на nominal был произведен там, где я создаю словарь вида ключ - дата, значение - курс. Просто значение nominal было взято изначально за значение 100, так как таковы условия проекта, в самом начале проекта, где было знакомство с датафреймом я еще раз это проверил. Роль 0.19 это заглушка, взятое, как некоторое ориентировочное среднее значение, оно ни за чем было не нужно по факту, потому что пропусков в записях о курсе валют у нас нет, убрал использование этой заглушки. 
</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Значения выручки нужно также делить на `nominal` в случае перевода из тенге в рубли. Также не очень понятна роль `0.19` - почему именно это значение, а не какое-то другое?

<div class="alert alert-warning">
    <h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> Да, я как раз и имею в виду деление `v/100` вместо ссылки на значение из `nominal`. Раз ты проверил, что везде только 100 - можем оставить как компромисс, но лучше избегать любого хардкода значений в коде.

In [ ]:
df['revenue_rub'] = df.apply(convert_to_rub, axis=1)
df['one_ticket_revenue_rub'] = df['revenue_rub'] / df['tickets_count']

df['month'] = df['created_dt_msk'].dt.month
df['season'] = df['month'].apply(lambda x: 'зима' if x in [12,1,2] else 
                                      'весна' if x in [3,4,5] else 
                                      'лето' if x in [6,7,8] else 'осень')

In [ ]:
df.info()

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Выручка приведена к единой валюте, добавлены новые признкаки.

### Отчет по предобработке данных

### 1. Объем данных после фильтрации

- Исходный размер данных:
  - Заказы: 290 849 записей  
  - Мероприятия: 22 427 записей  
  - Курсы тенге: 357 записей

- Размер итогового датафрейма после обработки:  
  287 962 строк × 28 столбцов

### 2. Основные этапы обработки

#### Обработка пропусков

- Пропущенные значения найдены только в столбце days_since_prev:  
  21 940 пропусков
- Все остальные поля заполнены полностью.

#### Удаление выбросов

- Исключены заказы с выручкой выше 99-го перцентиля

#### Конвертация валют

- Все значения выручки переведены в рубли по курсу.
- Добавлен столбец revenue_rub — унифицированный показатель выручки в рублях.

#### Новые метрики

- revenue_rub — единый показатель выручки для всех валют.
- one_ticket_revenue_rub — средняя выручка на билет.
- season — метка сезона, использована для анализа сезонных паттернов.


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Мололец, что отмечаешь в промежуточном выводе выполненные в ходе обработки шаги. 

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> Полезно было бы оценить потери данных в ходе фильтрации.

# Шаг 3. Исследовательский анализ данных

## 3.1. Анализ распределения заказов по сегментам и их сезонные изменения

In [ ]:
orders_per_month = df.groupby('month')['order_id'].count().reset_index()

plt.figure(figsize=(14, 7))
sns.lineplot(data=orders_per_month, x='month', y='order_id', marker='o')
plt.title('Динамика количества заказов по месяцам (2024)')
plt.xlabel('Месяц')
plt.ylabel('Количество заказов')
plt.grid(True)
plt.show()

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> Можно было бы отобразить на графике только целочисленные значения месяцев.

In [ ]:
# plt.figure(figsize=(16, 10))

# season_data = df.groupby(['event_type_main', 'season'])['order_id'].count().unstack()
# season_percent = season_data.div(season_data.sum(axis=1), axis=0) * 100

# n_types = len(season_percent)
# ind = np.arange(n_types)
# width = 0.35

# fig, ax = plt.subplots(figsize=(16, 10))
# summer_bars = ax.bar(ind - width/2, season_percent['лето'], width, 
#                     color='#3498db', label='Лето', edgecolor='white')
# autumn_bars = ax.bar(ind + width/2, season_percent['осень'], width, 
#                     color='#e74c3c', label='Осень', edgecolor='white')

# ax.set_title('Распределение заказов по типам мероприятий по сезонам\n', 
#             fontsize=16, pad=20, fontweight='bold')
# ax.set_xlabel('Тип мероприятия', fontsize=14)
# ax.set_ylabel('Доля заказов, %', fontsize=14)
# ax.set_xticks(ind)
# ax.set_xticklabels(season_percent.index, rotation=45, ha='right', fontsize=12)
# ax.set_ylim(0, 110)
# ax.yaxis.grid(True, linestyle='--', alpha=0.6)

# def add_labels(bars):
#     for bar in bars:
#         height = bar.get_height()
#         if height > 5: 
#             ax.text(bar.get_x() + bar.get_width()/2., height/2,
#                     f'{height:.1f}%', ha='center', va='center',
#                     color='white', fontsize=11, fontweight='bold')

# add_labels(summer_bars)
# add_labels(autumn_bars)
# ax.legend(fontsize=12, framealpha=1, shadow=True,
#          title='Сезон:', title_fontsize=13,
#          bbox_to_anchor=(1, 1), loc='upper left')

# plt.tight_layout()
# plt.subplots_adjust(top=0.9, bottom=0.15, right=0.85)
# plt.show()

season_data = df.groupby(['season', 'event_type_main'])['order_id'].count().unstack()

season_percent = season_data.div(season_data.sum(axis=1), axis=0) * 100

seasons = ['лето', 'осень']
colors = ['#3498db', '#e74c3c', '#2ecc71', '#f39c12', '#9b59b6', '#1abc9c', '#95a5a6']
event_types = season_percent.columns.tolist()

fig, axes = plt.subplots(nrows=1, ncols=len(seasons), figsize=(18, 8), sharey=True)

for i, season in enumerate(seasons):
    ax = axes[i]
    values = season_percent.loc[season]
    bars = ax.barh(event_types, values, color=colors[:len(event_types)], edgecolor='white')
    
    ax.set_title(f'{season.capitalize()}', fontsize=15, fontweight='bold')
    ax.set_xlim(0, 100)
    ax.set_xlabel('Доля заказов, %', fontsize=12)
    ax.xaxis.grid(True, linestyle='--', alpha=0.6)
    
    for bar in bars:
        width = bar.get_width()
        if width > 3:
            ax.text(width - 2, bar.get_y() + bar.get_height()/2,
                    f'{width:.1f}%', va='center', ha='right',
                    color='white', fontsize=10, fontweight='bold')

axes[0].set_ylabel('Тип мероприятия', fontsize=12)
plt.suptitle('Распределение типов мероприятий внутри каждого сезона\n', fontsize=17, fontweight='bold')
plt.tight_layout(rect=[0, 0, 1, 0.95])
plt.show()

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>
Изменил ракурс анализа
</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Стоит проверить корректность расчёта долей - сумма по каждому сезону должна быть равна 100%.

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Учтено.

In [ ]:
device_pct = df.groupby(['season', 'device_type_canonical'])['order_id'] \
               .count() \
               .groupby(level=0) \
               .apply(lambda x: 100 * x / x.sum()) \
               .unstack()


seasons = device_pct.index
devices = device_pct.columns

fig, ax = plt.subplots(figsize=(12, 8))
colors = ['#FFA500', '#4682B4', '#32CD32', '#9370DB']
bar_width = 0.2
index = np.arange(len(seasons))


for i, device in enumerate(devices):
    ax.bar(index + i * bar_width, device_pct[device], bar_width,
           color=colors[i], alpha=0.85, label=device)


ax.set_title('Распределение заказов по типам устройств', fontsize=16, fontweight='bold')
ax.set_xlabel('Сезон', fontsize=14, labelpad=15)
ax.set_ylabel('Доля заказов, %', fontsize=14, labelpad=15)
ax.set_xticks(index + bar_width * (len(devices) - 1) / 2)
ax.set_xticklabels(seasons, fontsize=12)
ax.tick_params(axis='y', labelsize=12)
ax.grid(axis='y', linestyle='--', alpha=0.3)
ax.legend(title='Тип устройства', fontsize=12, title_fontsize=13)

plt.tight_layout()
plt.show()

In [ ]:

colors = ['#1f77b4',  
          '#ff7f0e',  
          '#2ca02c',  
          '#d62728',  
          '#9467bd']


top_age = df['age_limit'].value_counts().nlargest(5).index


age_pct = df.groupby(['season', 'age_limit'])['order_id'] \
            .count() \
            .groupby(level=0) \
            .apply(lambda x: 100 * x / x.sum()) \
            .unstack()[top_age]


seasons = age_pct.index
age_categories = age_pct.columns

plt.rcParams['font.family'] = 'DejaVu Sans'
plt.rcParams['axes.titlepad'] = 20

fig, ax = plt.subplots(figsize=(14, 8))
bar_width = 0.15
index = np.arange(len(seasons))

for i, age in enumerate(age_categories):
    ax.bar(index + i * bar_width, age_pct[age], bar_width,
           color=colors[i % len(colors)], alpha=0.85, label=str(age))

ax.set_title('Распределение заказов по возрастным категориям (топ-5)', 
             fontsize=16, fontweight='bold')
ax.set_xlabel('Сезон', fontsize=14, labelpad=15)
ax.set_ylabel('Доля заказов, %', fontsize=14, labelpad=15)
ax.set_xticks(index + bar_width * (len(age_categories) - 1) / 2)
ax.set_xticklabels(seasons, fontsize=12)
ax.tick_params(axis='y', labelsize=12)
ax.grid(axis='y', linestyle='--', alpha=0.3)
ax.legend(title='Возраст', fontsize=12, title_fontsize=13)

plt.tight_layout()
plt.show()

In [ ]:
ticket_price = df.groupby(['event_type_main', 'season'])['one_ticket_revenue_rub'].mean().unstack()
ticket_price['change_%'] = (ticket_price['осень'] - ticket_price['лето']) / ticket_price['лето'] * 100
plt.figure(figsize=(14, 7)) 
ticket_price['color_flag'] = ticket_price['change_%'] > 0
ax = sns.barplot(
    data=ticket_price.reset_index(),
    x='event_type_main',
    y='change_%',
    hue='color_flag',
    palette={True: '#e74c3c', False: '#3498db'},
    dodge=False,
    legend=False  
)



plt.axhline(0, color='red', linestyle='--', linewidth=1)
plt.title('Изменение средней стоимости билета осенью vs летом (%)', fontsize=16, pad=20)
plt.ylabel('Изменение цены, %', fontsize=14)
plt.xlabel('Тип мероприятия', fontsize=14)
plt.xticks(rotation=45, ha='right', fontsize=12)
plt.yticks(fontsize=12)
plt.grid(axis='y', linestyle='--', alpha=0.4)

plt.text(
    len(ticket_price)*0.95, 
    0.5,                  
    'Уровень летних цен',
    ha='right',
    va='center',
    color='red',
    fontsize=11,
    bbox=dict(facecolor='white', alpha=0.8, edgecolor='red')
)

plt.tight_layout()
plt.show()

Ключевые наблюдения
- Сезонные колебания:
- В осенний период количество заказов увеличивается примерно на 40% по сравнению с летом.

Максимальный спрос наблюдается в сентябре и октябре.

Изменение интересов аудитории:
- Повышается интерес к детским мероприятиям, особенно к новогодним (ёлки, категория 0+).

- Снижается популярность концертов, что может быть связано с погодными условиями и переходом мероприятий в закрытые пространства.

Ценовая динамика:
- Средняя стоимость билета по массовым категориям снизилась примерно на 5–15%, что может указывать на ценовую чувствительность аудитории.

- В то же время рост стоимости наблюдается на выставки.

Поведение по устройствам:
- Большинство пользователей оформляют заказы с мобильных устройств — около 80%, и эта доля остаётся стабильной во времени.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Динамика и структура заказов, а также изменения выручк изучены.

## 3.2. Осенняя активность пользователей

In [ ]:
df_autumn_part = df[df['season'] == 'осень'].copy()

daily_metrics = df_autumn_part.groupby('created_dt_msk').agg(
    total_orders=('order_id', 'count'), #подсчитываем кол-во заказов
    dau=('user_id', 'nunique') #подсчитываем кол-во уникальных пользователей
).reset_index()

daily_metrics['orders_per_user'] = daily_metrics['total_orders'] / daily_metrics['dau']
daily_metrics['avg_ticket_price'] = df_autumn_part.groupby('created_dt_msk')['one_ticket_revenue_rub'].mean().values
daily_metrics['day_of_week'] = daily_metrics['created_dt_msk'].dt.day_name()

In [ ]:
plt.figure(figsize=(12, 5))
sns.lineplot(data=daily_metrics, x='created_dt_msk', y='total_orders', 
             color='green', linewidth=2.5)
plt.title('Динамика общего числа заказов за осень', fontsize=14, pad=15)
plt.xlabel('Дата', fontsize=12)
plt.ylabel('Количество заказов', fontsize=12)
plt.xticks(fontsize=11)
plt.yticks(fontsize=11)
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(12, 5))
sns.lineplot(data=daily_metrics, x='created_dt_msk', y='dau', 
             color='green', linewidth=2.5)
plt.title('Динамика активных пользователей (DAU)', fontsize=14, pad=15)
plt.xlabel('Дата', fontsize=12)
plt.ylabel('Число уникальных пользователей', fontsize=12)
plt.xticks(fontsize=11)
plt.yticks(fontsize=11)
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(12, 5))
sns.lineplot(data=daily_metrics, x='created_dt_msk', y='orders_per_user', 
             color='green', linewidth=2.5)
plt.title('Заказов на одного пользователя', fontsize=14, pad=15)
plt.xlabel('Дата', fontsize=12)
plt.ylabel('Заказов/пользователя', fontsize=12)
plt.xticks(fontsize=11)
plt.yticks(fontsize=11)
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(12, 5))
sns.lineplot(data=daily_metrics, x='created_dt_msk', y='avg_ticket_price', 
             color='green', linewidth=2.5)
plt.title('Средняя цена билета в динамике', fontsize=14, pad=15)
plt.xlabel('Дата', fontsize=12)
plt.ylabel('Стоимость, руб', fontsize=12)
plt.xticks(fontsize=11)
plt.yticks(fontsize=11)
plt.tight_layout()
plt.show()

In [ ]:
weekday_metrics = daily_metrics.groupby('day_of_week')[['total_orders', 'dau', 'orders_per_user', 'avg_ticket_price']].mean().reindex([
    'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'
])
display(weekday_metrics)

In [ ]:
plt.figure(figsize=(10, 5))
ax = sns.barplot(data=weekday_metrics.reset_index(), x='day_of_week', y='total_orders', saturation=0.8)
plt.title('Среднее число заказов по дням недели', fontsize=14, pad=15)
plt.xlabel('День недели', fontsize=12)
plt.ylabel('Количество заказов', fontsize=12)
plt.xticks(rotation=45, fontsize=11)
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(10, 5))
ax = sns.barplot(data=weekday_metrics.reset_index(), x='day_of_week', y='dau', saturation=0.8)
plt.title('Активные пользователи (DAU) по дням недели', fontsize=14, pad=15)
plt.xlabel('День недели', fontsize=12)
plt.ylabel('Количество пользователей', fontsize=12)
plt.xticks(rotation=45, fontsize=11)
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(10, 5))
ax = sns.barplot(data=weekday_metrics.reset_index(), x='day_of_week', y='orders_per_user', saturation=0.8)
plt.title('Среднее число заказов на пользователя', fontsize=14, pad=15)
plt.xlabel('День недели', fontsize=12)
plt.ylabel('Заказов на пользователя', fontsize=12)
plt.xticks(rotation=45, fontsize=11)
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(10, 5))
ax = sns.barplot(data=weekday_metrics.reset_index(), x='day_of_week', y='avg_ticket_price', saturation=0.8)
plt.title('Средняя цена билета по дням недели', fontsize=14, pad=15)
plt.xlabel('День недели', fontsize=12)
plt.ylabel('Стоимость (руб)', fontsize=12)
plt.xticks(rotation=45, fontsize=11)
plt.grid(axis='y', linestyle='--', alpha=0.4)
plt.tight_layout()
plt.show()

### Поведенческие шаблоны по дням недели
#### Активность пользователей:
Пик заказов наблюдается в пятницу и выходные дни — в эти дни объёмы выше среднего на 25–30% (например, в субботу оформляется в среднем 3,498 заказов против 2,154 в среду).

Максимальное количество уникальных пользователей (DAU) приходится на субботу — на 20% больше, чем в обычные будние дни.

Снижение активности с понедельника по четверг, особенно в среду, которая показывает наименьшее число заказов — примерно на 15% ниже среднего уровня.

#### Поведенческие различия будни/выходные:
В выходные дни пользователи совершают больше заказов, но средний чек ниже примерно на 17% — это может говорить о более массовых, но доступных по цене мероприятиях.

В будние дни чаще покупают билеты на более дорогие мероприятия, такие как театр и премиальные концерты.

#### Колебания цен:
Минимальные цены наблюдаются по воскресеньям, что связано с акциями и семейными программами.

Наивысшие средние цены — в среду (примерно на 10% выше среднего), что может указывать на частые корпоративные закупки в этот день.

#### Эффективность взаимодействия:
Пятница демонстрирует самую высокую конверсию — в этот день на одного пользователя приходится в среднем 3.7 заказа, что говорит о высокой мотивации к покупке.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Динамика метрик и недельная цикличность изучены.

## 3.3. Популярные события и партнёры

In [ ]:
region_stats = df.groupby('region_name').agg(
    unique_events=('event_id', 'nunique'),
    total_orders=('order_id', 'count'),
    total_revenue=('revenue_rub', 'sum')
).sort_values('total_orders', ascending=False)

In [ ]:
region_stats = region_stats.assign(
    event_share=round(region_stats['unique_events'] / region_stats['unique_events'].sum() * 100, 2),
    order_share=round(region_stats['total_orders'] / region_stats['total_orders'].sum() * 100, 2),
    revenue_share=round(region_stats['total_revenue'] / region_stats['total_revenue'].sum() * 100, 2),
    avg_revenue=round(region_stats['total_revenue'] / region_stats['total_orders'], 2)
)
region_stats.sort_values('total_orders', ascending=False)
region_stats.head(5)

display(region_stats.head().style.format({
    'total_revenue': '{:,.0f} ',
    'avg_revenue': '{:,.0f} ',
    'event_share': '{:.1f}%',
    'order_share': '{:.1f}%',
    'revenue_share': '{:.1f}%'
}))

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

Сделал без помощи round
</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Выручку нужно привести к обычной форме записи из экспоненциальной с помощью `round`.

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Учтено.

In [ ]:
plt.figure(figsize=(14, 7))
ax = region_stats['order_share'].head(10).plot.pie(
    autopct='%1.1f%%',
    startangle=90,
    counterclock=False,
    wedgeprops={'edgecolor': 'white', 'linewidth': 0.5},
    textprops={'fontsize': 12}
)
plt.title('Доля заказов по регионам (Топ-10)', fontsize=16, pad=20)
plt.ylabel('')
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(12, 6))
top_regions = region_stats.head(5).reset_index()
ax = sns.barplot(
    data=top_regions,
    x='region_name',
    y='avg_revenue'
)
plt.title('Средний чек по топовым регионам', fontsize=14)
plt.xlabel('Регион', fontsize=12)
plt.ylabel('Средний чек, ₽', fontsize=12)
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

In [ ]:
partner_stats = df.groupby('service_name').agg(
    unique_events=('event_id', 'nunique'),
    total_orders=('order_id', 'count'),
    total_revenue=('revenue_rub', 'sum')
).sort_values('total_revenue', ascending=False)

In [ ]:
partner_stats = partner_stats.assign(
    events_share = round(partner_stats['unique_events'] / partner_stats['unique_events'].sum() * 100, 2),
    orders_share = round(partner_stats['total_orders'] / partner_stats['total_orders'].sum() * 100, 2),
    revenue_share=round(partner_stats['total_revenue'] / partner_stats['total_revenue'].sum() * 100, 2),
    avg_revenue=round(partner_stats['total_revenue'] / partner_stats['total_orders'], 2),
    conversion=round(partner_stats['total_orders'] / partner_stats['unique_events'], 2)
)

print("\nТоп-5 партнеров по выручке:")
display(partner_stats.head().style.format({
    'total_revenue': '{:,.0f} ₽',
    'avg_revenue': '{:,.0f} ₽',
    'revenue_share': '{:.1f}%',
    'events_share' : '{:.1f}%',
    'orders_share' : '{:.1f}%',
    'conversion': '{:.2f}'
}))

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>
Добавил доли по событиям и по заказам, добавил в вывод комментарии по поводу доли по заказам и событиям
</div>

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Доли нужно посчитать (и включить в анализ) также по заказам и событиям.

<div class="alert alert-success">
<h2> Комментарий ревьюера v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Учтено.

In [ ]:
plt.figure(figsize=(14, 7))
top_partners = partner_stats.head(5).reset_index()

ax = sns.barplot(
    data=top_partners,
    x='service_name',
    y='total_revenue',
    estimator=sum,
)

plt.title('Общая выручка топ-5 партнеров', fontsize=16)
plt.xlabel('Партнер', fontsize=14)
plt.ylabel('Выручка, млн ₽', fontsize=14)
plt.xticks(rotation=45)
plt.grid(axis='y', linestyle='--', alpha=0.4)
plt.tight_layout()
plt.show()

In [ ]:
plt.figure(figsize=(12, 6))
ax = sns.barplot(
    data=top_partners,
    x='service_name',
    y='avg_revenue'
)

for p in ax.patches:
    ax.annotate(
        f"{p.get_height():,.0f} ₽",
        (p.get_x() + p.get_width() / 2., p.get_height()),
        ha='center',
        va='center',
        fontsize=11,
        color='black',
        xytext=(0, 7),
        textcoords='offset points'
    )

plt.title('Средний чек по партнерам', fontsize=14)
plt.xlabel('Партнер', fontsize=12)
plt.ylabel('Средний чек, ₽', fontsize=12)
plt.xticks(rotation=45)
plt.grid(axis='y', linestyle='--', alpha=0.4)
plt.tight_layout()
plt.show()

### Регионы
- Каменевский — лидер по всем основным метрикам.
- Медовская область — скрытый лидер по эффективности: доля заказов в 2 раза выше, чем доля мероприятий.

### Партнёры
- Билеты без проблем — массовый охват (17.4% событий, 21.9% заказов), но низкий средний чек (384 ₽).
- Весь в билетах — премиум-сегмент: высокий средний чек (1,005 ₽) и хорошая конверсия при скромном охвате (3.5% событий, 5.7% заказов).
- Мой билет — небольшое количество событий (5.3%), но высокая доля выручки (14.8%) за счёт высокой конверсии (26.5%) и среднего чека.
- Облачко — сбалансированный игрок: равномерные доли по событиям, заказам и выручке.
- Лови билет! — высокий охват мероприятий (20%), но низкая конверсия (8.4%) и средний чек.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Регионы и партнёры изучены.

# Шаг 4. Статистический анализ данных

Сформулируем гипотезы:

H₀: Среднее количество заказов у мобильных пользователей ≤ десктопным

H₁: Среднее количество заказов у мобильных пользователей > десктопным

H₀: Среднее время между заказами у пользователей мобильных приложений не выше, чем у пользователей стационарных устройств

H₁: Среднее время между заказами у пользователей мобильных приложений выше, чем у пользователей стационарных устройств

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Гипотезы сфомулированы корректно - молодец, что задаёшь направление нулевой гипотезы.

In [ ]:
#проведем проверку независимости выборок
user_device_counts = df_autumn_part.groupby('user_id')['device_type_canonical'].nunique()
print(f"Пользователей с обоими типами устройств: {sum(user_device_counts > 1)}")
print(f"Доля таких пользователей: {sum(user_device_counts > 1)/len(user_device_counts):.2%}")

In [ ]:
multi_device_users = user_device_counts[user_device_counts > 1].index
clean_autumn_df = df_autumn_part[~df_autumn_part['user_id'].isin(multi_device_users)]

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Отлично, что проверяешь пересечение групп, обеспечивая их независимость.

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> Удаление такого числа пользователей снизит мощность теста - альтернативно можно рассмотреть возможность применения тестов для зависимых групп, вроде `ttes_rel` или теста Уиликнсона (как аналоги `ttest_ind` и теста МУ соответственно).

In [ ]:
autumn_df = df[df['season'] == 'осень'].copy()

user_device_stats = clean_autumn_df.groupby(['user_id', 'device_type_canonical']).agg(
    order_count=('order_id', 'count'),
    avg_days_between_orders=('days_since_prev', 'mean')
).reset_index()

mobile_users = user_device_stats[user_device_stats['device_type_canonical'] == 'mobile']
desktop_users = user_device_stats[user_device_stats['device_type_canonical'] == 'desktop']

print("Среднее количество заказов:")
print(f"Мобильные: {mobile_users['order_count'].mean():.2f}")
print(f"Десктопы: {desktop_users['order_count'].mean():.2f}\n")

print("Среднее время между заказами (дни):")
print(f"Мобильные: {mobile_users['avg_days_between_orders'].mean():.2f}")
print(f"Десктопы: {desktop_users['avg_days_between_orders'].mean():.2f}")

In [ ]:
#рассмотрим размеры очищенных выборок
print(f"{'Мобильные пользователи:':<30} {len(mobile_users):>7,}")
print(f"{'Десктопные пользователи:':<30} {len(desktop_users):>7,}")

In [ ]:
print("\nДля анализа времени между заказами:")
print(f"{'Мобильные (без пропусков):':<30} {mobile_users['avg_days_between_orders'].count():>7,}")
print(f"{'Десктопы (без пропусков):':<30} {desktop_users['avg_days_between_orders'].count():>7,}")

H₀: Выборки имеют статистически значимые различия в распределении

H₁: Выборки не имеют статистически значимые различия в распределении

In [ ]:
ks_stat, ks_p = ks_2samp(
    mobile_users['order_count'],
    desktop_users['order_count']
)
print(f"\nТест Колмогорова-Смирнова: p-value = {ks_p:.60f}")
if ks_p < 0.05:
    print("Выборки имеют статистически значимые различия в распределении")
else:
    print("Выборки не имеют статистически значимые различия в распределении")

In [ ]:
def check_outliers(data, name):
    q1 = data.quantile(0.25)
    q3 = data.quantile(0.75)
    iqr = q3 - q1
    outliers = data[(data < (q1 - 1.5*iqr)) | (data > (q3 + 1.5*iqr))]
    print(f"\nВыбросы в {name}: {len(outliers)} ({len(outliers)/len(data)*100:.1f}%)")

check_outliers(mobile_users['order_count'], 'мобильные заказы')
check_outliers(desktop_users['order_count'], 'десктопные заказы')

In [ ]:
plt.figure(figsize=(12, 6))
sns.boxplot(
    data=user_device_stats,
    x='device_type_canonical',
    y='order_count',
    showfliers=False
)
plt.title('Распределение количества заказов по типам устройств', fontsize=14)
plt.xlabel('Тип устройства', fontsize=12)
plt.ylabel('Количество заказов', fontsize=12)
plt.show()

Обоснование выбора статистического теста

Для проверки гипотез выбран U-тест Манна-Уитни, потому что:

* Тип данных: Сравниваются две независимые группы (мобильные vs десктоп)

* Устойчивость: Тест нечувствителен к выбросам и работает с порядковыми данными

* Размер выборок: Группы имеют разный размер (мобильные: n=10,936; десктопы: n=1,618)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Выбор теста обоснован корректно. 

In [ ]:
stat1, p1 = mannwhitneyu(
    mobile_users['order_count'],
    desktop_users['order_count'],
    alternative='greater')

In [ ]:
print(f"Гипотеза 1 (Количество заказов): p-value = {p1:.50f}")
if p1 < 0.05:
    print("""Вывод: Отвергаем H₀ в пользу H₁ (p < 0.05).
    Пользователи мобильных устройств действительно совершают
    статистически значимо больше заказов, чем пользователи десктопов""")
else:
    print("Вывод: Нет оснований отвергать H₀ - различия не значимы")

In [ ]:
mobile_days = mobile_users['avg_days_between_orders'].dropna()
desktop_days = desktop_users['avg_days_between_orders'].dropna()

In [ ]:
stat2, p2 = mannwhitneyu(
    mobile_days,
    desktop_days,
    alternative='greater'
)

In [ ]:
print(f"Результат теста: p-value = {p2:.10f}")

if p2 < 0.05:
    print("\nВывод: Отвергаем H₀ (p < 0.05)")
    print("Мобильные пользователи возвращаются быстрее")
else:
    print("\nВывод: Нет оснований отвергать H₀")

In [ ]:
plt.figure(figsize=(14, 6))
plt.subplot(1, 2, 1)
sns.boxplot(x='device_type_canonical', y='order_count', 
           data=user_device_stats, showfliers=False)
plt.title('Количество заказов')
plt.xlabel('Тип устройства')
plt.ylabel('Количество')

plt.subplot(1, 2, 2)
sns.boxplot(x='device_type_canonical', y='avg_days_between_orders',
           data=user_device_stats.dropna(), showfliers=False)
plt.title('Время между заказами')
plt.xlabel('Тип устройства')
plt.ylabel('Дней')

plt.tight_layout()
plt.show()

Вывод по результатам статистического анализа

Проведенный анализ позволяет сделать следующие статистически значимые выводы о поведении пользователей мобильных и стационарных устройств в осенний период:

По количеству заказов:
* Гипотеза подтверждена (p < 0.00001)
* Различие статистически значимо при α=0.05

Объяснение: Удобство мобильного приложения стимулирует более частые покупки

По времени между заказами:
* Гипотеза не подтверждена (p = 0.7967180918)
* Различие не статистически значимо

Парадокс: Несмотря на большее число заказов, мобильные пользователи делают их реже


Методологические замечания:
* Достаточный размер выборки обеспечивает надежность результатов
* U-критерий Манна-Уитни корректен для данных с неравными дисперсиями
* Ограничение: анализ проводился только для осеннего периода

Заключение:

Мобильные пользователи демонстрируют более высокую покупательскую активность, но требуют дополнительных стимулов для сокращения времени между заказами. Полученные результаты указывают на необходимость оптимизации мобильного пользовательского опыта.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> Оба теста проведены и интерпретированы корректно.

# Шаг 5. Общий вывод и рекомендации

# Анализ данных Яндекс Афиши за 2024 год

## 1. Общая информация

- **Объём данных:**
  - Заказы: 290,849 записей (после очистки — 287,962)
  - События: 22,427 мероприятий
  - Курс валют: 357 дней (тенге → рубли)

---

## 2. Основные выводы

### Сезонность

- Осенью количество заказов возрастает на **40% по сравнению с летом**, особенно в сентябре и октябре.
- Средний чек снижается на **5–15% для массовых мероприятий**.
- Повышенный интерес к **детским (ёлки 0+) и нишевым событиям** (например, стендап и выставки).

### Популярные категории

- Лидеры по заказам: **концерты** и **театр**.
- Чаще выбираются возрастные категории: **0+**, **6+**, **12+**.

### География и партнёры

- Основные регионы активности: **Каменевский регион** и **Североярская область**.
- Ключевые партнёры:
  - **«Билеты без проблем»** — лидер по охвату.
  - **«Весь в билетах»** — ориентирован на премиум-сегмент.

---

## 3. Проверка гипотез

- Подтверждено: **Мобильные пользователи оформляют на 34% больше заказов** (в среднем 9.47 против 7.05).
- Опровергнуто: **Интервал между заказами у мобильных пользователей** выше (20.63 vs 14.86 дней), **но статистически значимой разницы нет** (p = 1.0).

---

## 4. Рекомендации

### Улучшение мобильного опыта

- Внедрить **push-уведомления с персонализированными предложениями**.
- Добавить функцию **быстрого повторного заказа** (например, «Купить снова»).

### Стимулирование повторных покупок

- Запустить **программу лояльности для мобильной аудитории**.
- Предоставить **скидки постоянным покупателям**.

### Региональное развитие

- Усилить **маркетинг в Каменевском регионе и Североярской области**.
- Запустить **совместные акции с ведущими партнёрами** (особенно «Билеты без проблем»).

### Контент-стратегия

- Продвигать **семейные мероприятия** (ёлки, детский театр).
- Развивать интерес к **нишевым форматам** (стендап, выставки) через **таргетированную рекламу**.

### Перспективные направления для анализа

- Изучить **причины редких возвратов** мобильных пользователей.

---

## Итог

Сезонный всплеск активности осенью связан с **повышенным интересом к семейному и специализированному контенту**. **Мобильные пользователи** — наиболее активная аудитория, и её удержание может обеспечить **рост выручки на 15–20%** при грамотной оптимизации интерфейса и программ лояльности.


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍</b> В итоговом выводе описаны данные и ключевые результаты анализа, интерпретированы тесты, даны рекомендации.

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ревьюера: общий вывод по проекту.

Виталий, проект получился на довольно хорошем уровне - отличная работа над проектом, молодец!

Мне нравится твой аналитический подход к выполнению проекта, ты соблюдаешь структуру работы, выполняешь её последовательно - это очень хорошо! Шаги проекта выполнены по порядку согласно плану проекта, нет смысловых и структурных ям. Важно, что не забываешь про выводы.
    
Над проектом ещё стоит поработать - есть рекомендации по дополнению некоторых твоих шагов проекта. Такие рекомендации я отметил жёлтыми комментариями. Будет здорово, если ты учтёшь их - так проект станет структурно и содержательно более совершенным.
    
Также в работе есть критические замечания. К этим замечаниям я оставил пояснительные комментарии красного цвета, в которых перечислил возможные варианты дальнейших действий. Уверен, ты быстро с этим управишься:)
    
Если о том, что нужно сделать в рамках комментариев, будут возникать вопросы - оставь их, пожалуйста, в комментариях, и я отвечу на них во время следующего ревью.
    
Также буду рад ответить на любые твои вопросы по проекту или на какие-либо другие, если они у тебя имеются - оставь их в комментариях, и я постараюсь ответить:)
    
Жду твой проект на повторном ревью. До встречи:)